tpm2tools/keys.go

```diff
@@ -133,3 +134,173 @@ func (k *Key) PublicKey() crypto.PublicKey {
+}
+
+// Seal seals the sensitive byte buffer to the provided PCRs under the owner
+// hierarchy using the SHA256 versions of the provided PCRs.
+// The Key k is used as the parent key.
+func (k *Key) Seal(rw io.ReadWriter, pcrs []int, sensitive []byte) (SealedBytes, error) {
+	auth, err := getPCRSessionAuth(rw, pcrs)
+	if err != nil {
+		return SealedBytes{}, fmt.Errorf("could not get pcr session auth: %v", err)
+	}
+
+	return sealHelper(rw, k.Handle(), auth, sensitive)
+}
+
+func sealHelper(rw io.ReadWriter, parentHandle tpmutil.Handle, auth []byte, sensitive []byte) (SealedBytes, error) {
+	priv, pub, err := tpm2.Seal(rw, parentHandle, "", "", auth, sensitive)
+	if err != nil {
+		return SealedBytes{}, fmt.Errorf("failed to seal data: %v", err)
+	}
+	return SealedBytes{priv, pub}, nil
+}
+
+// Unseal takes a private/public pair of buffers and attempts to reverse the
+// sealing process under the owner hierarchy using the SHA256 versions of the
+// provided PCRs.
+// The Key k is used as the parent key.
+func (k *Key) Unseal(rw io.ReadWriter, pcrs []int, in SealedBytes) ([]byte, error) {
+	session, err := createPCRSession(rw, pcrs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unseal: %v", err)
+	}
+	defer tpm2.FlushContext(rw, session)
+
+	sealed, _, err := tpm2.Load(
+		rw,
+		k.Handle(),
+		/*parentPassword=*/ "",
+		in.pub,
+		in.priv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load sealed object: %v", err)
+	}
+	defer tpm2.FlushContext(rw, sealed)
+
+	return tpm2.UnsealWithSession(rw, session, sealed, "")
+}
+
+// Reseal unwraps a secret and rewraps it under the auth value that would be
+// produced by the PCR state in pcrs. Similar to seal and unseal, this acts on
+// the SHA256 PCRs and uses the ownerhierarchy.
+// The Key k is used as the parent key.
+func (k *Key) Reseal(rw io.ReadWriter, pcrs map[int][]byte, in SealedBytes) (SealedBytes, error) {
+	pcrNums := make([]int, 0, len(pcrs))
+	for key := range pcrs {
+		pcrNums = append(pcrNums, key)
+	}
+
+	sensitive, err := k.Unseal(rw, pcrNums, in)
+	if err != nil {
+		return SealedBytes{}, fmt.Errorf("failed to unseal: %v", err)
+	}
+
+	auth, err := computePCRSessionAuth(pcrs)
+	if err != nil {
+		return SealedBytes{}, fmt.Errorf("failed to compute pcr session auth: %v", err)
+	}
+
+	return sealHelper(rw, k.Handle(), auth, sensitive)
+}
+
+type tpmsPCRSelection struct {
+	Hash tpm2.Algorithm
+	Size byte
+	PCRs tpmutil.RawBytes
+}
+
+type sessionSummary struct {
+	OldDigest      tpmutil.RawBytes
+	CmdIDPolicyPCR uint32
+	NumPcrSels     uint32
+	Sel            tpmsPCRSelection
+	PcrDigest      tpmutil.RawBytes
+}
+
+func computePCRSessionAuth(pcrs map[int][]byte) ([]byte, error) {
+	var pcrBits [3]byte
+	for pcr := range pcrs {
+		byteNum := pcr / 8
+		bytePos := byte(1 << byte(pcr%8))
+		pcrBits[byteNum] |= bytePos
+	}
+	pcrDigest := digestPCRList(pcrs)
+
+	summary := sessionSummary{
+		OldDigest:      make([]byte, sha256.Size),
+		CmdIDPolicyPCR: 0x0000017f,
```

Comment:
https://github.com/google/go-tpm/pull/67 has been merged, so this can be changed.

Reply:
Done.